CASSANDRA-19777: Update doc page for how-to commit with more intro (#311)
apache/cassandra-website: site-content/source/modules/ROOT/pages/development/how_to_commit.adoc

Reviewer: ```suggestion
Fix land on the oldest release branch, and are then forward-merged onto each newer branch using an ours merge strategy to record branch lineage.  Each forward-merge commit contains the branch-appropriate patch.
```

Commits applicable to multiple versions are atomically pushed forward merges.

> Fix land on the oldest release branch, and are then forward-merged onto each newer branch using an ours merge strategy to record branch lineage.  Each forward-merge commit contains the branch-appropriate patch.

Reviewer: sry, one more attempt… 
```suggestion
Fixes are applied first on the oldest applicable release branch, and are then forward-merged onto each newer branch using an ours merge strategy to record branch lineage.  Each forward-merge commit contains the branch-appropriate patch.
```

This keeps a clean, traceable history and a single logical unit of work per ticket per branch, while preventing unintended diffs from being pulled forward automatically.

== Introduction for New Committers

> To be an effective committer, you should be familiar with the following git command lines. The following outlines the simple trunk-only patch workflow.  See next section for multi-branch contributions.

Reviewer: ```suggestion
Patches are applied using git command lines, not via github UI.  The following outlines the simple trunk-only patch workflow.  See next section for multi-branch contributions.
```

> Patches are applied using git command lines, not via github UI.  The following outlines the simple trunk-only patch workflow.  See next section for multi-branch contributions.

Reviewer 2: Suggest:

GitHub Pull Requests are merged using Git command-line tools instead of the GitHub UI. The following workflow outlines a standard trunk-only patching process; for contributions involving multiple branches, see the next section.

> Development branches are kept in forks.  The upstream apache/cassandra repository is reserved for trunk and release branches.

Reviewer 2: Development occurs within personal forks, as the upstream ``apache/cassandra`` repository is strictly reserved for the trunk and official release branches.

> * **Local feature branches:** Always do your work in a local feature/topic branch. Do not push your personal feature branches upstream; only the canonical trunk and release branches should be pushed to the shared repository when merging changes.

Reviewer: ```suggestion
Patches are applied using git command lines, not via github UI.  The following outlines the simple trunk-only patch workflow.  See next section for multi-branch contributions.

Development branches are kept in forks.  The upstream apache/cassandra repository is reserved for trunk and release branches.
```

Use the following Git CLI commands when merging a PR:

Step 1: From your project repository, check out a new branch and test the changes.
[source,shell]
----
git checkout -b <jira>/<branch> <branch>
----

Step 2: Squash, fast-forward merge, and update on GitHub following the project workflow.
[source,shell]
----
git switch trunk
# there should only be one commit, squashed 

> git cherry-pick <jira>/<branch>

Reviewer 2: cherry pick a branch or commit-hash?

git push --atomic origin trunk -n

> # check dry-run looks correct

Reviewer 2: shouldn't the dry run comment (-n) be before the command not after?  I realize it's suggesting to check afterwards, but might be be more clear if the dry run was explained before executed.

``# push as dry-run (-n) and check that it looks correct``

> # push as dry-run (-n) and check that it looks correct

Reviewer 2: Think one needs to update commit message in correct form (as discussed later in doc)

`git commit --amend`

> git push --atomic origin trunk -n

Reviewer 2: Should it be upstream (github.com/apache/cassandra) not origin (your clone)? In which case, Step 1 needs to set upstream and verify 

```
git remote add upstream https://gitbox.apache.org/repos/asf/cassandra.git
git remote -v
```

Commits applicable to multiple versions are atomically pushed forward merges.

Fix land on the oldest release branch, and are then forward-merged onto each newer branch using an ours merge strategy to record branch lineage.  Each forward-merge commit contains the branch-appropriate patch.

This keeps a clean, traceable history and a single logical unit of work per ticket per branch, while preventing unintended diffs from being pulled forward automatically.

== Introduction for New Committers

To be an effective committer, you should be familiar with the following background details. While this is not a git tutorial, it outlines the expected workflow:

> * **Using CLI vs GitHub web UI:** We recommend using the Git CLI rather than the GitHub web UI.

Reviewer: ```suggestion
Patches are applied using git command lines, not via github UI.  The following outlines the simple trunk-only patch workflow.  See next section for multi-branch contributions.

```

This keeps a clean, traceable history and a single logical unit of work per ticket per branch, while preventing unintended diffs from being pulled forward automatically.

== Introduction for New Committers

To be an effective committer, you should be familiar with the following background details. While this is not a git tutorial, it outlines the expected workflow:

> * **Using CLI vs GitHub web UI:** We recommend using the Git CLI rather than the GitHub web UI.

Reviewer 2: but why?

* **Local feature branches:** Always do your work in a local branch, and never push a branch upstream.
* **Workflow:** This repository uses a squash-rebase-merge workflow with feature branches.
* **Basic git commands you should know:**
** `git remote add` & `git remote set-url`
** Squashing commits with `git rebase -i`
** Updating a commit with `git commit --amend`
** `git push upstream <repo> --atomic -n`

> GitHub offers these CLI commands for merging a PR:

Reviewer 2: Should be git commands, not github

Step 1: From your project repository, check out a new branch and test the changes.
[source,shell]
----
git checkout -b omniCoder77-feature/guardrail-for-misprepare-statements trunk

> git pull https://github.com/omniCoder77/cassandra.git feature/guardrail-for-misprepare-statements

Reviewer 2: It's general practice to use branch names with the Jira ID in the name, e.g.  CASSANDRA-9999. You can look at some open PRs for examples.

This keeps a clean, traceable history and a single logical unit of work per ticket per branch, while preventing unintended diffs from being pulled forward automatically.

== Introduction for New Committers

To be an effective committer, you should be familiar with the following background details. While this is not a git tutorial, it outlines the expected workflow:

* **Using CLI vs GitHub web UI:** We recommend using the Git CLI rather than the GitHub web UI.
* **Local feature branches:** Always do your work in a local feature/topic branch. Do not push your personal feature branches upstream; only the canonical trunk and release branches should be pushed to the shared repository when merging changes.

> * **Workflow:** This repository uses a squash-rebase-merge workflow with feature branches. Fixes land on the oldest supported branch and are merged forward into newer branches (using ours merges and amended merge commits).

Reviewer: remove.  not applicable to this section.
```suggestion
```

* **Basic git commands you should know:**
** `git remote add` & `git remote set-url`
** Squashing commits with `git rebase -i`
** Updating a commit with `git commit --amend`

> ** `git push origin <branches> --atomic -n`

Reviewer: remove, no added value here
```suggestion
```

Use the following Git CLI commands when merging a PR:

Step 1: From your project repository, check out a new branch and test the changes.
[source,shell]
----
git checkout -b <github_username>-feature/<pr_branch> trunk
git fetch https://github.com/<github_username>/<repo_name>.git <pr_branch>:<github_username>-feature/<pr_branch>

> git checkout <github_username>-feature/<pr_branch>

Reviewer: ```suggestion
git checkout -b <jira>/<branch> <branch>
```

----

Step 2: Squash, fast-forward merge, and update on GitHub following the project workflow.
[source,shell]
----
git rebase -i trunk   # squash to a single, well-titled commit
git checkout trunk
git merge --ff-only <github_username>-feature/<pr_branch>

> git push origin trunk --atomic

Reviewer: ```suggestion
git switch trunk
# there should only be one commit, squashed 
git cherry-pick <jira>/<branch>
git push --atomic origin trunk -n
# check dry-run looks correct
git push --atomic origin trunk
```
